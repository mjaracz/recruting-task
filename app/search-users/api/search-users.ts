import { pageRoutes, apiRoutes } from "./routes";

export interface User {
  username: string;
  avatarUrl: string;
}

export interface SearchUsersData {
  totalItems: number;
  items: User[];
  page: number;
  perPage: number;
}

export const getSearchUser = async (
  page: string,
  search: string,
): Promise<SearchUsersData> => {
  let prepredUrl = `${pageRoutes.main}/${apiRoutes.getSearchUser}?${new URLSearchParams(
    {
      page: page,
      per_page: "10",
    },
  )}`;
  if (search) {
    prepredUrl += `&q=${search}`;
  }

  if (search)
    return fetch(prepredUrl)
      .then((res) => {
        if (res.status <= 299) return res.json();
        else throw new Error(`API exception with status code: ${res.status}`);
      })
      .then((data) => ({
        totalItems: data.total_count,
        items: data.items.map(
          (item: { login: string; avatar_url: string }) => ({
            username: item.login,
            avatarUrl: item.avatar_url,
          }),
        ),
        page: parseInt(page),
        perPage: 10,
      }));
  else
    return new Promise<SearchUsersData>((res) =>
      res({ totalItems: -1, items: [], page: -1, perPage: -1 }),
    );
};
