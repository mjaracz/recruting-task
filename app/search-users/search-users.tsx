"use client";

import { ListItem } from "@/components/list-item";
import { TextField } from "@/components/text-field";
import { Title } from "@/components/title";
import { CircularProgress, List } from "@mui/material";
import { default as MaterialListItem } from "@mui/material/ListItem";
import { Snackbar } from "@/components/snackbar";
import { useSearchUsers } from "./hooks/useSearchUsers";

export const SearchUsers = () => {
  const {
    onChangeSearch,
    isLoading,
    users,
    error,
    sentryRef,
    hasNextPage,
    isErrorOpen,
  } = useSearchUsers();

  return (
    <>
      <Title label="Search Github Users" />
      <TextField onChange={onChangeSearch} variant="outlined" />

      <div>
        {isLoading && <CircularProgress size={80} />}
        {!!users.length && !error && (
          <List
            sx={{
              display: "flex",
              width: "100%",
              maxWidth: 360,
              bgcolor: "background.paper",
              justifyContent: "center",
              flexDirection: "column",
              alignItems: "center",
              padding: "2rem",
              marginTop: 5,
            }}
          >
            {users.map((user, index) => (
              <ListItem key={`${user.username}_${index}`} {...user} />
            ))}
            {(isLoading || hasNextPage) && (
              <MaterialListItem
                ref={sentryRef}
                sx={{
                  display: "flex",
                  justifyContent: "center",
                  alignItems: "center",
                  padding: "10px",
                }}
              >
                <CircularProgress size={80} />
              </MaterialListItem>
            )}
          </List>
        )}
        <Snackbar
          alertStatus="error"
          message={error?.message}
          open={isErrorOpen}
        />
      </div>
    </>
  );
};
