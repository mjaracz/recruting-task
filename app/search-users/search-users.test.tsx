import { render } from '@testing-library/react';
import { SearchUsers } from "./search-users";
import Providers from '../providers';

describe('<SearchUsers />', () => {
  test('should correct match the snapshot', () => {
    const { baseElement } = render(<Providers><SearchUsers/></Providers>);

    expect(baseElement).toMatchSnapshot();
  })
})