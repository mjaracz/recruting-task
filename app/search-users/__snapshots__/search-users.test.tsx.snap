// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchUsers /> should correct match the snapshot 1`] = `
<body>
  <div>
    <h2
      class="title"
    >
      Search Github Users
    </h2>
    <div
      class="MuiFormControl-root MuiTextField-root TextField css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1rh62vq-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-rg9sn3-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-gpjuor"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div>
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-yknyv-MuiCircularProgress-root"
        role="progressbar"
        style="width: 80px; height: 80px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-1a8epmm-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;
